wazuh/wazuh-kibana-app
# Centralize the plugin settings

# Description

This PR:
- Add:
  - Define the plugin setting schema.
    - Adapt the current logic to use the plugin schema
  - Enhance the plugin setting description displayed in the UI and the configuration file.
  - Add description for the plugin settings categories
- Change `PUT /utils/configuration` endpoint to support multiple settings
- Internal
  - Add:
    - New hook to manage the state of form with changes.
  - Change:
    - Refactor the components related to the input form in `Settings/Configuration`
- Test
  - Add tests for:
    - Backend endpoints:
      - `GET /utils/configuration`
      - `PUT /utils/configuration`
      - `GET /utils/logs`
   - Default configuration file
   - Methods related to the new plugin setting definition

Closes #4443 #4450

# Tests

## Manual tests

**Scenario**: The configuration file is generated with the expected settings.
  **When** the plugin platform starts and there is no plugin configuration file in the expected path
  **Then** a new configuration file is generated and this should contain all the expected settings
  
**Scenario**:  The configuration through `Settings/Configuration` section works
  **When** the user go to `Settings/Configuration`
  **Then** change the input values works as expected
 
 > This scenario applies to each setting. Test each setting.
 
 **Scenario**: The user saves the configuration through `Settings/Configuration`
   **Given**: the user changes 1 or more settings
   **When**: the user clicks on the `Save` button
   **Then**: the configuration file should be updated with the changed settings and the frontend should update its values
   
**Scenario**: The action toast when saving the configuration can only show one each time.
  **Given**: the user changes 2 or more settings that require the same type of action after updating.
  **When**: the user clicks on the `Save` button
  **Then**: it should only appear a toast showing the action to do.
  
  > This scenario applies to each type of current actions: Execute healtcheck, reload the browser tab and restart the plugin platform
  
**Scenario**: Filter settings by category in `Settings/Configuration`
  **When** the user selects one or more categories from select input on top of the page.
  **Then** the settings of the selected category/ies should be displayed.
  
**Scenario**: User search a setting in `Settings/Configuration`
  **When** the user selects introduces a query in the search bar.
  **Then** the setting/s that match the query are displayed.
  
**Scenario**: Update one setting from `Settings/Configuration`
  **When** the user changes a setting value and clicks on the `Save` button
  **Then** the setting should be updated in the configuration file and the frontend
  
**Scenario**: Update 2 or more settings from `Settings/Configuration`
  **When** the user changes 2 or more setting values and clicks on the `Save` button
  **Then** the settings should be updated in the configuration file and the frontend
  
**Scenario**: Update one setting through API request `PUT /utils/configuration`
  **When** the user does a request to update some setting using some request tool: cURL, postman, etc...
  **Then** the setting should be updated in the configuration file
  
**Scenario**: Update 2 or more settings through API request `PUT /utils/configuration`
  **When** the user does a request to update 2 or more settings using some request tool: cURL, postman, etc...
  **Then** the settings should be updated in the configuration file  
  
**Scenario**: Plugin setting category description appears in the configuration file
  **Given** an environment without a plugin configuration file
  **When** the user starts the plugin platform and the default configuration file is generated
  **Then** each plugin category has the description on the configuration file
  
**Scenario**: Plugin setting category description appears in the `Settings/Configuration`
  **When** the user navigates to `Settings/Configuration`
  **Then** each plugin category has the description 
  
  ***
  To manually test the configuration endpoint here are 2 scripts to update different configuration values.
  
[centralize-conf.zip](https://github.com/wazuh/wazuh-kibana-app/files/9734914/centralize-conf.zip)


To use it, first unzip the files in the same folder and allow both _executing the file as a program_. Afterwards you can run the `manual-test-centralize.sh` script.

You can also use **wz-api-conf.sh** script with custom configuration parameters, like:
```
\- s setting_name
\- v setting_value
```

eg.
`./wz-api-conf.sh -s disabled_roles -v [\"disabled\"]`

#### Output

```
$ ./manual-test-centralize.sh 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    839      0 --:--:-- --:--:-- --:--:--   837
{"timeout":25000}
{"data":{"requiresRunningHealthCheck":false,"requiresReloadingBrowserTab":false,"requiresRestartingPluginPlatform":false,"updatedConfiguration":{"timeout":25000}}}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    914      0 --:--:-- --:--:-- --:--:--   914
{"extensions.aws":true}
{"data":{"requiresRunningHealthCheck":false,"requiresReloadingBrowserTab":false,"requiresRestartingPluginPlatform":false,"updatedConfiguration":{"extensions.aws":true}}}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    855      0 --:--:-- --:--:-- --:--:--   855
{"alerts.sample.prefix":"prefix_test_"}
{"data":{"requiresRunningHealthCheck":true,"requiresReloadingBrowserTab":false,"requiresRestartingPluginPlatform":false,"updatedConfiguration":{"alerts.sample.prefix":"prefix_test_"}}}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    904      0 --:--:-- --:--:-- --:--:--   902
{"ip.ignore":["test*\"]}
{"statusCode":400,"error":"Bad Request","message":"Invalid request payload JSON format"}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    898      0 --:--:-- --:--:-- --:--:--   898
{"logs.level":"debug"}
{"data":{"requiresRunningHealthCheck":false,"requiresReloadingBrowserTab":false,"requiresRestartingPluginPlatform":true,"updatedConfiguration":{"logs.level":"debug"}}}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    910      0 --:--:-- --:--:-- --:--:--   910
{"fake.config":"test fake config"}
{"statusCode":400,"error":"Bad Request","message":"[request body.fake.config]: definition for this key is missing"}
```


## Automatic tests
- `server/routes/wazuh-utils/wazuh-utils.test.ts`
- `public/components/common/form/hooks.test.tsx`
- `public/components/common/form/index.test.tsx`
- `server/lib/initial-wazuh-config.test.ts`
- `common/services/settings.test.ts`

Run these tests with:
```
yarn test:jest <filepath>
```

**Run all PR tests**
```shell
yarn test:jest server/routes/wazuh-utils/wazuh-utils.test.ts public/components/common/form/hooks.test.tsx public/components/common/form/index.test.tsx server/lib/initial-wazuh-config.test.ts common/services/settings.test.ts
```
